pkg/web/apps/apps_psql_repository_test.go

```diff
@@ -1,56 +1,241 @@
-// +build integration
-
 package apps
 
 import (
-	"reflect"
 	"testing"
 
-	"github.com/aerogear/mobile-security-service/pkg/config"
-	"github.com/aerogear/mobile-security-service/pkg/db"
-
 	"github.com/aerogear/mobile-security-service/pkg/models"
+	sqlmock "gopkg.in/DATA-DOG/go-sqlmock.v1"
 )
 
-func Test_appsPostgreSQLRepository_GetApps(t *testing.T) {
-	c := config.Get()
-	dbConn, err := db.Connect(c.DB.ConnectionString, c.DB.MaxConnections)
+func Test_appsPostgreSQLRepository_GetAppsWillReturnTwoApps(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		t.Fatalf("Unexpected error opening a stub database connection: %v", err)
+	}
+
+	defer db.Close()
+
+	mockApps := []models.App{
+		models.App{
+			ID:      "7f89ce49-a736-459e-9110-e52d049fc025",
+			AppID:   "com.aerogear.mobile_app_one",
+			AppName: "Mobile App One",
+		},
+		models.App{
+			ID:      "7f89ce49-a736-459e-9110-e52d049fc026",
+			AppID:   "com.aerogear.mobile_app_three",
+			AppName: "Mobile App Two",
+		},
+		models.App{
+			ID:      "7f89ce49-a736-459e-9110-e52d049fc027",
+			AppID:   "com.aerogear.mobile_app_three",
+			AppName: "Mobile App Three",
+		},
+	}
+
+	cols := []string{"id", "app_id", "app_name", "deleted_at"}
+
+	timestamp := "2019-02-15T09:38:33+00:00"
+
+	// Insert an app where the deleted_at column is set
+	sqlmock.NewRows(cols).AddRow(mockApps[0].ID, mockApps[0].AppID, mockApps[0].AppName, timestamp)
+
+	// Insert 2 apps which are not soft deleted
+	rows := sqlmock.NewRows(cols).AddRow(mockApps[1].ID, mockApps[1].AppID, mockApps[1].AppName, "").AddRow(mockApps[2].ID, mockApps[2].AppID, mockApps[2].AppName, "")
+
+	query := `
```

Comment:
Why have 2 places with the same implementation/duplicate code? It shows that should be centralized.

Reply:
This is required for [go-sqlmock](https://github.com/DATA-DOG/go-sqlmock). It compare this SQL query to the one in the actual repository function and fails if they are mismatched.

[Example](https://github.com/bxcodec/go-clean-arch/blob/master/article/repository/mysqlarticle_test.go#L59)

Reply:
The point here is that we should centralize the code in one place. In the way that it is implemented if be required change the query then need to look for many places since the same query/string/code can be found in more than 2 places. Is it make sense?

Reply:
@camilamacedo86 I partially agree with you - there is a duplication of content here. However this would make our unit tests dependent on an external state which is not good practice either. I am not sure of the best approach here.

Also, notice that the query string is different in the test file. This is because this is is a regex and not a pure SQL query string. It does a compare to the actual query being performed.

I've taken a step to reduce duplication by having the regex strings to the top of the test file.

Reply:
HI @craicoverflow,

Following my IHMO. If the same code is duplicated many times in this file only. Then a simple solution can add a global const var with this info that can be used just in this class/routine.

Reply:
Hi @camilamacedo86 this is what I suggested above:

> I've taken a step to reduce duplication by having the regex strings to the top of the test file.

Reply:
Closing it as resolved so.

Reply:
HI @craicoverflow,

Just to complement and share. If we are using the same query in more than one routine we could centralize it doing something like the following example.



```
type QueryStore struct {
    db          *sql.DB
    prepared    map[string]*sql.Stmt   // you might want to use an Enum type instead of string for added safety
}

func NewQueryStore(db *sql.DB) *QueryStore {
    q := &QueryStore{db: db, prepared: map[string]*sql.Stmt{}}
    
    // add your prepared statements to the map
    stmt, err := db.Prepare("bar")
    checkErr(err)
    q.prepared[foo] = stmt

    // etc

    return q
}

func (s QueryStore) getNames(param string) []string{
    rows, err := s.prepared[foo].Query(param)
    // error checking
    var results []string
    // row scanning
    // more error checking
    // etc...
    return results
} 
```